File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/RevokeKeyCmd.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import picocli.CommandLine;
import sop.Ready;
import sop.cli.picocli.SopCLI;
import sop.exception.SOPGPException;
import sop.operation.RevokeKey;

import java.io.IOException;

@CommandLine.Command(name = "revoke-key",
        resourceBundle = "msg_revoke-key",
        exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
public class RevokeKeyCmd extends AbstractSopCmd {

    @CommandLine.Option(names = "--no-armor",
            negatable = true)
    boolean armor = true;

    @CommandLine.Option(names = "--with-key-password",
            paramLabel = "PASSWORD")
    String withKeyPassword;

    @Override
    public void run() {
        RevokeKey revokeKey = throwIfUnsupportedSubcommand(
                SopCLI.getSop().revokeKey(), "revoke-key");

        if (!armor) {
            revokeKey.noArmor();
        }

        if (withKeyPassword != null) {
            try {
                String password = stringFromInputStream(getInput(withKeyPassword));
                revokeKey.withKeyPassword(password);
            } catch (SOPGPException.UnsupportedOption e) {
                String errorMsg = getMsg("sop.error.feature_support.option_not_supported", "--with-key-password");
                throw new SOPGPException.UnsupportedOption(errorMsg, e);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        Ready ready;
        try {
            ready = revokeKey.keys(System.in);
        } catch (SOPGPException.KeyIsProtected e) {
            String errorMsg = getMsg("sop.error.runtime.cannot_unlock_key", "STANDARD_IN");
            throw new SOPGPException.KeyIsProtected(errorMsg, e);
        }
        try {
            ready.writeTo(System.out);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/RevokeKeyCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import java.io.IOException
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import sop.cli.picocli.SopCLI
import sop.exception.SOPGPException
import sop.exception.SOPGPException.KeyIsProtected

@Command(
    name = "revoke-key",
    resourceBundle = "msg_revoke-key",
    exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
class RevokeKeyCmd : AbstractSopCmd() {

    @Option(names = ["--no-armor"], negatable = true) var armor = true

    @Option(names = ["--with-key-password"], paramLabel = "PASSWORD", arity = "0..*")
    var withKeyPassword: List<String> = listOf()

    override fun run() {
        val revokeKey = throwIfUnsupportedSubcommand(SopCLI.getSop().revokeKey(), "revoke-key")

        if (!armor) {
            revokeKey.noArmor()
        }

        for (passwordIn in withKeyPassword) {
            try {
                val password = stringFromInputStream(getInput(passwordIn))
                revokeKey.withKeyPassword(password)
            } catch (e: SOPGPException.UnsupportedOption) {
                val errorMsg =
                    getMsg("sop.error.feature_support.option_not_supported", "--with-key-password")
                throw SOPGPException.UnsupportedOption(errorMsg, e)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }

        val ready =
            try {
                revokeKey.keys(System.`in`)
            } catch (e: KeyIsProtected) {
                val errorMsg = getMsg("sop.error.runtime.cannot_unlock_key", "STANDARD_IN")
                throw KeyIsProtected(errorMsg, e)
            }
        try {
            ready.writeTo(System.out)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }
}
